### DISCUSSION TITLE
[GSoC]: DoQ x DoH3

### DISCUSSION THREAD

**Discussion by jaehnri** [coredns/coredns#6000] [Q&A]
Hi, everyone!
I've talked to Chris and I want to participate in GSoC 2023. I propose to add support to either DoQ or DoH3. 
I've been studying the related RFCs (QUIC, DoQ, DoH) and how to implement DoQ these last few days and I'm unsure whether I should progress with DoQ or DoH3. I've listed some pros and cons below.

DoH3:
- Pros
    - [Caddy supports HTTP/3](https://github.com/caddyserver/caddy)
    - It already has [cool use cases](https://security.googleblog.com/2022/07/dns-over-http3-in-android.html), mainly for mobile devices.
- Cons
    - HTTP/3 is never mentioned in the [DoH RFC](https://www.rfc-editor.org/rfc/rfc8484) and I think DoH3 isn't an IETF defined standard yet.
    - Go's standard library [does not support HTTP/3](https://github.com/golang/go/issues/32204). yet.

DoQ:
- Pros
    - DoQ has its [own RFC](https://datatracker.ietf.org/doc/rfc9250/).
- Cons 
    - Caddy does not directly support QUIC. gRPC is also "not supported" but DNS-over-gRPC is implemented anyway.
    - Go's standard library also does not support QUIC, but [quic-go](https://datatracker.ietf.org/doc/rfc9250/) looks reliable enough.

From what I've seen, DoH3 could be harder to implement but could be potentially more useful as some Android phones are already using it. However, I think the path to implementing DoQ is much more straightforward and could be easier to do with the current machinery (quic-go). 

I know that we could and should implement both, but I would like to choose one for the GSoC proposal. Surely, after one is implemented, the other should be easier to do too. I would love to hear your thoughts here. Thanks in advance.

**Comment by jaehnri** [comment]
cc @chrisohaver @yongtang @Tantalor93 @miekg

**Comment by Tantalor93** [comment]
Hello, I think starting with the DoH/3 might the best, especially considering the use case by Android, so we know we will be writing support for something that could be useful for someone

**Reply by jaehnri** [reply]
I agree. One question: as there already is the `forward` protocol, should I make it support DoH/3 or create a new plugin similar to [gRPC's](https://github.com/coredns/coredns/tree/master/plugin/grpc)? I saw Miek [comment a few times](https://github.com/coredns/coredns/issues/3788#issuecomment-606415567) that it should be a new plugin, so I think I'll proceed it this way...

**Reply by Tantalor93** [reply]
hmm, I think I am kinda confused now in what manner we want to support DoH/3  or DoQ. Do we want  CoreDNS to support accepting these protocols as server, or do we want to support CoreDNS communicating with other resolvers using these protocols? These are quite different problems with different solutions, which one do we want to tackle? I thought we are talking about the former, but it seems you are referencing issues regarding adding support to forward plugin

**Reply by jaehnri** [reply]
The idea is to support both, much like gRPC does today ([server_grpc.go](https://github.com/coredns/coredns/blob/master/core/dnsserver/server_grpc.go) and [gRPC plugin](https://github.com/coredns/coredns/tree/master/plugin/grpc)):
1) We'll set up a QUIC server so CoreDNS can serve QUIC requests;
2) Create a new QUIC plugin for CoreDNS to communicate to upstream QUIC resolvers using DoQ;

I have described both of these in my proposal. I'll send it to you via e-mail. Comments and suggestions are appreciated.

**Comment by yongtang** [comment]
@jaehnri  The GSoC 23 submission timeline already passed. We may have to contact CNCF to see if it is still possible to add additional project in after the timeline. On the other hand, we can also add the project to Linux Foundation's Mentorship program which will sponsor in a similar way like GSoC.

**Reply by jaehnri** [reply]
Hi, @yongtang!
Submissions for GSoC '23 are open until April 4. As CoreDNS has participated many times in the past under CNCF, I assumed that it wouldn't be a problem to submit this proposal, but I may be missing something. I'll check if we need to do something else apart from the proposal.

**Reply by chrisohaver** [reply]
I think Yong is saying that although submissions are open until 4th, the deadline for CoreDNS to apply as a participating project has passed.  My apologies, I am very ignorant of the GSoC process.

**Reply by jaehnri** [reply]
Oh, I see...
Well, I contacted Nate W., one of CNCF's responsible advocates, to see if CoreDNS is eligible and if we can do something about it. No problem if there's no way.

I browsed the #mentoring channel a little and [there was indeed a deadline for project ideas](https://github.com/cncf/mentoring/discussions/777). The same discussion also states that "all CNCF projects are eligible to apply to participate". However, ideas are still being submitted, [such as these from 3 days ago](https://cloud-native.slack.com/archives/CGPK98JNQ/p1680043931476979).

I guess that it is possible to submit proposals that aren't on the project ideas list, but I have no clue how they select them. If I get any updates from Nate I'll send them here.

**Reply by jaehnri** [reply]
OK, so here's Nate's answer: https://cloud-native.slack.com/archives/CGPK98JNQ/p1680289545603739

TLDR: 
- the project should be submitted to the [project ideas list](https://github.com/cncf/mentoring/blob/main/summerofcode/2023.md) first and then I can send my application. If this is possible to do, I would be very glad. If you folks need help, I can write the summary so one of you can open the PR.
- [LFX project proposals due](https://github.com/cncf/mentoring/tree/main/lfx-mentorship/2023/02-Jun-Aug) are also open until April 28th.

**Comment by yongtang** [comment]
I have opened a PR in https://github.com/cncf/mentoring/pull/911. @Tantalor93 @chrisohaver  let me know if you want to add your name to the mentor list, I will update the PR.

**Reply by jaehnri** [reply]
Thank you so much! I'll send the proposal PDF for you to review ASAP.
Should I send it in Slack or elsewhere?

**Reply by yongtang** [reply]
You can send through email if you want early comments. The final submission will need to be done through GSoC project page (if the proposal is accepted).

GSoC project is open to public so it might be possible to see multiple submissions to one project. In that case one submission will be accepted based on the quality of the project submissions.

**Reply by Tantalor93** [reply]
as I said in https://github.com/coredns/coredns/discussions/6000#discussioncomment-5506937, I think we should determine what kind of support we want to add and mention it in the PR:

1. CoreDNS accepting DoQ and DoH/3 as server
2. CoreDNS forwarding DNS requests using DoQ and DoH/3

**Reply by jaehnri** [reply]
I think it is okay if we don't mention both of these in the PR, but I've detailed in my proposal that I'll support both of the features above.

**Reply by chrisohaver** [reply]
@yongtang, please add me as a mentor, thanks!
